docs/chapters/incremental_indexing/index.md

```diff
@@ -0,0 +1,29 @@
+# Incremental Indexing in Jina
+
+## Generating Documents Ids:
+
+In Jina, the document ID is by default generated deterministically by computing a hash the document content. This avoids repeating the doc id while reindexing by using a Random or Incremental counter. The user can `override` document ids with Jina client and add `tags` to map to their unique concepts.
+
+## Incremental Indexing Design Pattern:
+
+The rationale behind Incremental indexing is to avoid re-indexing previously indexed documents, thus saving time and compute. This is done by persisting the document identifiers of indexed docs and using them as a DocIdCache.
+While indexing a document, it’s identifier is checked against the Cache for a ‘hit’ - if the document id exists,it’s already been indexed. We skip re-indexing it. If it’s a ‘miss’, we index it, adding the identifier to the Cache.
+1. DocIdCache stores doc ids in a int64 set and persists it to a numpy array
+2. UniqueVectorIndexer is a design pattern for combining VectorIndexer with DocIdCache
+3. UniquePbIndexer is a common design pattern for combining KVIndexer with DocIdCache
+
+The difference between a cache and a KVIndexer is the handler_mutex is released in cache. This allows a user to query-while-indexing. This is because we can support querying already indexed doc ids in the cache, since they are not re-indexed on hit. In the meantime, non-indexed doc ids get added to the cache during indexing.
```

Comment:
I would not comment the `query-while-indexing` because it gives a sense that it is a feature that we can now support

Reply:
sure, I removed it and documented that is a possibility in the future that we can support it